File: content/manual/manual/source/includes/csfle/ruby/make_data_key.rb
require 'mongo'
require 'base64'
require 'securerandom'
require_relative 'config'

# start-generate-cmk
File.binwrite(MASTER_KEY_PATH, SecureRandom.random_bytes(96))
# end-generate-cmk

def make_key
  # start-create-index
  key_vault_client = Mongo::Client.new(CONNECTION_STRING)
  key_vault_client.use(KEY_VAULT_DB).database.drop
  key_vault_client.use('medicalRecords').database.drop
  key_vault_client.use(KEY_VAULT_DB)[KEY_VAULT_COLL].indexes.create_one(
    { 'keyAltNames' => 1 },
    unique: true,
    partial_filter_expression: { 'keyAltNames' => { '$exists' => true } }
  )
  key_vault_client.close
  # end-create-index

  # start-create-data-key
  client = Mongo::Client.new(CONNECTION_STRING)
  client_encryption = Mongo::ClientEncryption.new(
    client,
    key_vault_namespace: KEY_VAULT_NAMESPACE,
    kms_providers: get_kms_providers
  )
  dek_id = client_encryption.create_data_key('local')
  base64_dek_id = Base64.strict_encode64(dek_id.data)
  puts "DataKeyId [base64]: #{base64_dek_id}"
  File.write(DEK_ID_PATH, base64_dek_id)
  client.close
  # end-create-data-key
end

File: content/manual/manual/source/includes/csfle/go/make_data_key.go
package main

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"fmt"
	"os"

	"go.mongodb.org/mongo-driver/v2/bson"
	"go.mongodb.org/mongo-driver/v2/mongo"
	"go.mongodb.org/mongo-driver/v2/mongo/options"
)

func MakeKey() error {
	// start-generate-cmk
	key := make([]byte, 96)
	if _, err := rand.Read(key); err != nil {
		return fmt.Errorf("could not generate random key: %v", err)
	}
	if err := os.WriteFile(masterKeyPath, key, 0644); err != nil {
		return fmt.Errorf("could not write master key to file: %v", err)
	}
	// end-generate-cmk

	// start-create-index
	keyVaultClient, err := mongo.Connect(options.Client().ApplyURI(connectionString))
	if err != nil {
		return fmt.Errorf("connect error for key vault client: %v", err)
	}
	defer func() {
		_ = keyVaultClient.Disconnect(context.TODO())
	}()

	// Drops the key vault collection and database if they already exist
	if err = keyVaultClient.Database(keyVaultDb).Collection(keyVaultColl).Drop(context.TODO()); err != nil {
		return fmt.Errorf("collection drop error: %v", err)
	}
	if err = keyVaultClient.Database("medicalRecords").Collection("patients").Drop(context.TODO()); err != nil {
		return fmt.Errorf("collection drop error: %v", err)
	}
	keyVaultIndex := mongo.IndexModel{
		Keys: bson.D{{Key: "keyAltNames", Value: 1}},
		Options: options.Index().
			SetUnique(true).
			SetPartialFilterExpression(bson.D{
				{Key: "keyAltNames", Value: bson.D{
					{Key: "$exists", Value: true},
				}},
			}),
	}
	_, err = keyVaultClient.Database(keyVaultDb).Collection(keyVaultColl).Indexes().CreateOne(context.TODO(), keyVaultIndex)
	if err != nil {
		return fmt.Errorf("index create error: %v", err)
	}
	// end-create-index

	// start-create-data-key
	kmsProviders, err := getKmsProviders()
	if err != nil {
		return err
	}
	clientEncryptionOpts := options.ClientEncryption().
		SetKeyVaultNamespace(keyVaultNamespace).
		SetKmsProviders(kmsProviders)
	clientEnc, err := mongo.NewClientEncryption(keyVaultClient, clientEncryptionOpts)
	if err != nil {
		return fmt.Errorf("NewClientEncryption error: %v", err)
	}
	defer func() {
		_ = clientEnc.Close(context.TODO())
	}()

	dataKeyID, err := clientEnc.CreateDataKey(context.TODO(), "local", options.DataKey())
	if err != nil {
		return fmt.Errorf("create data key error: %v", err)
	}

	base64DekId := base64.StdEncoding.EncodeToString(dataKeyID.Data)
	fmt.Printf("DataKeyId [base64]: %s\n", base64DekId)
	// Writes the key ID to a file for use by other files
	if err := os.WriteFile(dekIdPath, []byte(base64DekId), 0644); err != nil {
		return fmt.Errorf("write dek_id.txt error: %v", err)
	}
	// end-create-data-key

	return nil
}
